### PR TITLE
Wip/prototype

### PR DESCRIPTION
**Pull request by lorenzkuhn**
### Current Status
`bimodal_posterior.py` contains a first draft of BBBVI applied to the model we discussed in our last meeting: 
mu ~ N(0, 5)
sigma = 1
x ~ N(mu^2, sigma^2)

We use BBBVI to approximate the posterior over mu. Given the observations `data = torch.tensor([4.0, 4.2, 3.9, 4.1, 3.8, 3.5, 4.3])`, we expect the posterior over `mu` to be a bimodal distribution with modes around -2 and +2. With the right initialisation that is indeed what our prototype converges to. One can reproduce this by running `python3 bimodal_posterior.py`.

### Possible Next Steps

- Check whether my BBBVI implementation is correct.
- The output of the current algorithm varies a lot for different parameter initialisations and learning rates. We should probably test different initialisations in a more structured way and if possible find a method of finding good initialisations automatically. 
- So far we've just set `lambda = 1`, we'll have to modify RELBO to allow for different values of lambda. 
- Check whether the current implementation properly uses TraceEnum_ELBO.

### CODE REVIEW

**Comment by TNU-yaoy** [comment]
> We use BBBVI to approximate the posterior over mu. Given the observations `data = torch.tensor([4.0, 4.2, 3.9, 4.1, 3.8, 3.5, 4.3])`, we expect the posterior over `mu` to be a bimodal distribution with modes around -2 and +2. With the right initialisation that is indeed what our prototype converges to. One can reproduce this by running `python3 bimodal_posterior.py`.

Without having thought very deeply about this, I wonder if you could apply pyro's standard SVI (with  normal elbo loss) to our toy model with the same data to see if the initialization problems is specific to our extension to boosting or if our model is simply bad for convergence. Intuitively I would expect the standard VI to jump between the solutions +2 and -2 depending on initial values.

**Comment by lorenzkuhn** [comment]
Yeah, I think that makes sense and I would expect the same to happen. I'll give it a try👍🏼

**Comment by gideonite** [comment]
Looks good to me. Let's merge. I might make a PR later with some minor modifications.

Next steps from my view:
1. incorporate lambda --- just for consistency
2. add smarter mixture weight setting --- requires implementing our latest work, I can do this if that's helpful.
3. try Bayesian Logistic Regression --- if we want to automatically adjust hyperparams, I think we need to learn more about different situations.
     - or, we can think of another toy example
